Document the export CSV format

ggalmazor opened the pull request:
This PR adds a documentation file with information about the CSV format Briefcase uses to export submissions.

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+
+### Fixed block columns
+
+| Column name    | Optional | Quoted | Description |
```
For optional, I was expecting the inverse. That is, I want to know what things are Required.

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+| Separator               | `,`   | Comma character |
+| Quoted                  | `"`   | Double quote character. See the [Values block columns](#values-block-columns) section to know which fields get quoted |
+| Header                  | Yes   | Every file will have a header with the respective column names |
+| Empty columns           | Yes   | See the "Values block columns" section to know which fields get empty columns |
```
Add a link for Values block columns

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+
+### Structure of the file:
+
+| Column block | Description |
```
I would drop this table and consolidate the information in the next table.

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+| Column name    | Optional | Quoted | Description |
+| -------------- | -------- | ------ | ----------- |
+| SubmissionDate | No       | Yes    | Holds the submission date. See the [Values block columns](#values-block-columns) section to learn how it is encoded |
+| KEY            | No       | No     | Holds the `instanceID` of each submission |
```
I would add another row below SubmissionDate that represents the values rows.

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+|                        | Pattern                                                                            | Example                                               |
+| ---------------------- | ---------------------------------------------------------------------------------- | ----------------------------------------------------- |
+| Topmost level pattern  | `{INSTANCE ID}`                                                                    | `uuid:00000000-0000-0000-0000-000000000000`           |
+| Pattern of descendants | `{RECURSIVE PARENT KEY}/{PARENT GROUP NAME}[{PLACE OF PARENT AMONG ITS SIBLINGS}]` | `uuid:00000000-0000-0000-0000-000000000000/group1[1]` |
```
s/{PLACE OF PARENT AMONG ITS SIBLINGS}/{ORDERING}/g

yanokwa replied:
It's not obvious what RECURSIVE PARENT KEY means. Can you give me an example and maybe I could come up with a different wording?

ggalmazor replied:
Let's say that there is 3 nested groups g1, g2, and g3.

```
g1 PARENT KEY uuid:00000000-0000-0000-0000-000000000000
g2 PARENT KEY uuid:00000000-0000-0000-0000-000000000000/g1[1]
g3 PARENT KEY uuid:00000000-0000-0000-0000-000000000000/g1[1]/g2[1]
```

Do you think we need to change the examples on the doc's table?

ggalmazor replied:
Would `{PARENT'S PARENT KEY}` work? I know the word "recursive" can be daunting.

yanokwa replied:
What about `grandparent`. I think your existing examples are good enough.

ggalmazor replied:
I think you're right! I was concerned about  `GRANDPARENT KEY` maybe being misleading because maybe `PARENT KEY != KEY(parent)` but it turns out that it holds.

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+
+|                        | Pattern                                                                            | Example                                               |
+| ---------------------- | ---------------------------------------------------------------------------------- | ----------------------------------------------------- |
+| Topmost level pattern  | `{INSTANCE ID}`                                                                    | `uuid:00000000-0000-0000-0000-000000000000`           |
```
Maybe parent pattern and children pattern?

ggalmazor replied:
Same argument as before: a child can be a parent of some other field. I'm not sure about changing the semantics.

As an alternative, we could use "root" and "node", but I'd suggest changing the code to use the same language. These dissonances between the domain and the source code are confusing down the road.

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+| KEY            | No       | No     | Holds the `instanceID` of each submission |
+| isValidated    | Yes      | No     | It's only present if the form is encrypted. It holds `true` if the cryptographic signature has been validated, `false` otherwise |
+
+## Repeat output file
```
I wonder if it's clearer to Title this "Child output file" and reduce then mention repeat in the follow on paragraph?

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+
+### Date and time fields
+
+These fields get encoded following the user's locale settings and using the [MEDIUM style pattern](https://docs.oracle.com/javase/7/docs/api/java/text/DateFormat.html).
```
s/using the/Java's/

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+
+These fields get encoded following the user's locale settings and using the [MEDIUM style pattern](https://docs.oracle.com/javase/7/docs/api/java/text/DateFormat.html).
+
+### Non empty value codification
```
s/non empty/non-empty/g

ggalmazor commented on docs/export-format.md:
```diff
@@ -0,0 +1,158 @@
+| Empty columns           | Yes   | See the "Values block columns" section to know which fields get empty columns |
+| Encoding of null values | No    | Nulls are just empty columns and, sometimes, empty strings |
+
+## Main submission output file
```
Are you sure about this change? All throughout the code the concept that is used is "main" and "repeat" and parent might be misleading since a repeat group can be parent of a nested repeat group, and this second group can be also parent of a third nested group.

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,146 @@
+|                        | Pattern                                       | Example                                                      |
+| ---------------------- | --------------------------------------------- | ------------------------------------------------------------ |
+| Topmost level pattern  | `{INSTANCE ID}/{GROUP NAME}`                  | `uuid:00000000-0000-0000-0000-000000000000/group1`           |
+| Pattern of descendants | `{RECURSIVE SET-OF}[{ORDERING OF THE GROUP}]` | `uuid:00000000-0000-0000-0000-000000000000/group1[1]/group2` |
```
s/RECURSIVE/GRANDPARENT/

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,146 @@
+| Empty columns           | Yes   | See the [Values block columns](#values-block-columns) section to know which fields get empty columns |
+| Encoding of null values | No    | Nulls are empty columns and, sometimes, empty strings |
+
+## Main submission output file
```
s/submission//

yanokwa commented on docs/export-format.md:
```diff
@@ -0,0 +1,146 @@
+
+## Main submission output file
+
+The main submission out file has values from any non-repeating group field found at the topmost level of a form's model and other metadata.
```
s/submission out/output/

ggalmazor commented:
I've studied a little bit more the code that generates the keys for the repeat files and I've simplified it (I've also updated the export efficiency PR).

I've changed the first column of those tables to "Nesting Level" and rows are "First" and "Following". What do you think?

yanokwa commented:
Nesting level language feels like a step back. First and following don't have the same vertical hierarchy to me as Top and descendants. The latter also has a family vibe to it that first the parent language we have in the doc.

ggalmazor commented:
OK, I've fixed the language to top/descendant group. Let me know if there's something left ;)

ggalmazor commented:
@yanokwa, @lognaturel, @matthew-white, I'm sorry I forgot about this... I've updated the PR with Matt's feedback.

yanokwa commented:
I this is good enough to merge. @clint-tseng if in your work on Central you run into questions, please file an issue so we can fix!
